### ISSUE TITLE
Filesystem module: getMode vs getPermissions

### ISSUE THREAD

**Issue by jhh67** [chapel-lang/chapel#19656]
Several concerns regarding the current `getMode` method:

1. Should it be renamed to `getPermissions`? 
2. Should it return something other than an `int`, perhaps a record with fields for the individual mode bits.

**Comment by lydia-duncan** [comment]
Personally, I think getPermissions is too long a name and like getMode just fine.  I am intrigued by the idea of returning a record with fields for the mode bits, though I think it would be useful to still have the more performant/lighter weight int return available

**Comment by bradcray** [comment]
I think Vass's suggestion, which was intriguing, was to have it return a record that simply contained the `int` (say, `var mode: int;`) so that it would return a type with a more descriptive name than `int` and that type could support richer queries like `userCanWrite()` (or whatever we'd want to support) while still permitting the power user to get right at the int if they wanted to do bit operations on it or the like.  Downside: more stuff to name and define.  Upside: calls different things by different names, slightly more future-proof (?), and supports the ability to add more descriptive queries on the type rather than via a richer call.

**Comment by daviditen** [comment]
Answering some context questions that came up today from @mstrout:

> What is the current method name with an example of how it is used?

[getMode(name: string): int throws](https://chapel-lang.org/docs/master/modules/standard/FileSystem.html#FileSystem.getMode)

```chapel
const mode = getMode(filename1);
chmod(filename2, 644);
writeln("chmod works: ", getMode(filename2) == 644);
chmod(filename2, mode); // change it back
```

> How often is the current method used in CHAMPS and Arkouda and internally in Chapel code?
It is not used in CHAMPS or Arkouda. It is only used in a few spots in the FileSystem module and in functionality tests.

> What are the key points being made by people who would like to see a change before stabilization?

I think that mode can imply more than just permissions. For example in might include file type in addition.

> How does this functionality overlap with what is in OS.Posix?  Should we have the overlap or not?

There is quite a bit of overlap between them, they each declare constants for each permissions bit. It looks like the only thing in OS.Posix for getting the mode bits is to use [stat](https://chapel-lang.org/docs/master/modules/standard/OS/POSIX.html#POSIX.stat) and extract them from the struct it fills in.

> How does this functionality compare with similar Python functionality?
Python has [os.stat()](https://www.geeksforgeeks.org/python-os-stat-method/) similar to what's in our `OS.posix` module.

**Comment by bradcray** [comment]
I think in the meeting today it was suggested that whether we kept getMode() or getPermissions(), we'd probably keep the name `chmod()`?  Checking OS.POSIX, I'm seeing that it supports `chmod()` which suggests to me that `FileSystem` could (and probably should) support something more symmetrical with whatever this routine ends up being named (e.g., setMode()/changeMode() or setPermissions()/changePermissions()).

**Comment by bradcray** [comment]
Belated thought:  Can permission modes even be considered to be OS-neutral?  Aren't our notions of ugo+rwx very POSIX-specific where Windows (say) might have different permissions modes (a subset or superset)?  And if so, is the right approach here to put any routines that manipulate or reason about permissions into the OS.POSIX module, as chmod is, and then deprecate their equivalents in the FileSystem module?  (and if/when we have a non-POSIX set of routines that span capabilities present across multiple OSes, put those in OS?)

**Comment by daviditen** [comment]
OK, so the proposal then is to deprecate everything about permissions from `FileSystem` (`getMode`, `chmod`, `copyMode`, the `S_IRUSR` style constants) in favor of existing functionality in OS.Posix? Or would we want to add `getPermissions` and `copyPermissions` to Posix in addition to what is already there?

**Comment by bradcray** [comment]
> OK, so the proposal then is to deprecate everything about permissions from FileSystem (getMode, chmod, copyMode, the S_IRUSR style constants) in favor of existing functionality in OS.Posix? 

That sounds right to me.

> Or would we want to add getPermissions and copyPermissions to Posix in addition to what is already there?

Since the contents of POSIX are a pretty direct transliteration of the C interface, I think we'd want to support these capabilities in the usual POSIX way with the usual namings rather than these Chapeltastic ones.  It wouldn't be surprising if those features already existed, though, knowing Greg?  E.g., I think I understood yesterday that `stat()` is the POSIX way to getPermissions, and that seems to be there (though it's using a renaming that would be good to understand).  I'm not sure what the POSIX equivalent to copyPermissions would be, though...

Assuming they are there, I'd probably convert any useful tests we might have for those cases over to use the POSIX equivalents, unless we already have existing POSIX tests that are equivalent.

**Comment by bradcray** [comment]
I'd also be interested in getting @lydia-duncan to check this thinking due to her mention of how heavily this module had been designed yesterday (though in our defense, it was designed before OS.POSIX existed or we were thinking much about future non-POSIX support) and @mppf since he was pretty involved in the discussion yesterday as well.

**Comment by mppf** [comment]
I'm supportive of the plan. I'll restate it to make sure we are understanding each other:
 * deprecating getMode / copyMode / chmod in the FileSystem module
 * suggesting to users to use the equivalent POSIX functionality if they need it
 * in the future (presumably after Chapel 2.0), consider adding a portable and Chapel-tastic way to do this. Such an effort would need us to understand what the functions / "permissions" ideas would mean in a non-POSIX OS that we are supporting (e.g. Windows)

**Comment by daviditen** [comment]
I'd be good with that plan too.

**Comment by bradcray** [comment]
> though it's using a renaming that would be good to understand

Ah, looks like the funny renaming around stat() is due to some POSIX/Darwin mismatches in field names, so a little "huh..."-esque, but not a reason to not proceed with the plan.

**Comment by daviditen** [comment]
Closed with #21130
